modules/social_features/social_core/src/Plugin/Field/FieldType/ExternalIdentifierItem.php

```diff
@@ -0,0 +1,221 @@
+    foreach ($this->getValue() as $value) {
+      if ($value !== '' && $value !== NULL) {
+        return FALSE;
+      }
+    }
+
+    return TRUE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function propertyDefinitions(FieldStorageDefinitionInterface $field_definition): array {
+    $properties['external_id'] = DataDefinition::create('string')
+      ->setLabel(t('External ID'))
+      ->setRequired(TRUE);
+    $properties['external_owner_target_type'] = DataReferenceTargetDefinition::create('string')
+      ->setLabel(t('Target Entity Type'))
+      ->setRequired(TRUE);
+    // @todo See if we can improve this subfield to use
+    //   DataReferenceDefinition or DataDynamicReferenceDefinition.
+    $properties['external_owner_id'] = DataDefinition::create('integer')
+      ->setLabel(t('External Owner'))
+      ->setRequired(TRUE);
+
+    return $properties;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function storageSettingsSummary(FieldStorageDefinitionInterface $storage_definition): array {
+    $summary = [];
+    $target_types = $storage_definition->getSetting('target_types');
+    if (!empty($target_types)) {
+      $entity_type_labels = [];
+      foreach ($target_types as $target_type) {
+        $target_type_info = \Drupal::entityTypeManager()->getDefinition($target_type);
+        if (!empty($target_type_info)) {
+          $entity_type_labels[] = $target_type_info->getLabel();
+        }
+      }
+      $summary[] = new TranslatableMarkup('Reference type: @entity_type_labels', [
+        '@entity_type_labels' => implode(', ', $entity_type_labels),
+      ]);
+
+    }
+
+    return $summary;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function schema(FieldStorageDefinitionInterface $field_definition): array {
+    return [
+      'columns' => [
+        'external_id' => [
+          'type' => 'varchar',
+          'description' => 'External ID',
+          'length' => 255,
+          'not null' => TRUE,
+        ],
+        'external_owner_target_type' => [
+          'type' => 'varchar',
+          'description' => 'External Owner Target Type',
+          'length' => 255,
+          'not null' => TRUE,
+        ],
+        'external_owner_id' => [
+          'type' => 'int',
+          'description' => 'External Owner ID',
+          'unsigned' => TRUE,
+          'not null' => TRUE,
+        ],
+      ],
+      'indexes' => [
+        'external_id' => ['external_id'],
+        'external_owner_target_type' => ['external_owner_target_type'],
+        'external_owner_id' => ['external_owner_id'],
+      ],
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setValue($values, $notify = TRUE): void {
+    if (is_array($values)) {
+      $this->set('external_id', $values['external_id'] ?? NULL);
+      $this->set('external_owner_target_type', $values['external_owner_target_type'] ?? NULL);
+      $this->set('external_owner_id', $values['external_owner_id'] ?? NULL);
```

Comment:
I realise we have a constraint that checks that all these fields are set later. However, I wonder if from a developer experience point of view it wouldn't be better to throw an `InvalidArgumentException` if any of those fields are not set.

For example, would we find an implementation that only sets some of those fields a correct implementation, or would we consider that to be a bug by whoever is using this field?

Reply:
> I realise we have a constraint that checks that all these fields are set later.
Yes indeed

We could already throw `InvalidArgumentException` at `setValue` but IMHO constraints are IMHO more appropriate place to do it.

Reply:
I think it depends on what we want to achieve and how Drupal's constraint system fits into other data integrity decisions.

In my understanding, Drupal's constraint system is mostly there to help user input and runs only when `validate` is called on an entity. This means that programmatically we might still be able to create a field item that's invalid just because validation is bypassed. The question is then whether we want to allow this, or whether we want to prevent it and let a developer know "Hey you forgot to validate your inputs, that's not right!"

Throwing an `InvalidArgumentException` is one way to let a developer know their implementation is doing something wrong that they should change. So it depends on when `setValue` is called and where that exception ends up. If throwing an exception in `setValue` would mean that when a user submits an invalid form they no longer get a form validation error but instead the application crashes, then throwing an exception here is not the right answer. In that case we may want to think about "How do we warn a developer that they're trying to save an invalid field value because they (intentionally) skipped the validate step" and solve that problem.

Reply:
@Kingdutch , I just had a discussion with Zan where he explained that even if they are three different subfields, they come as a package, so we can't update just one part of that field, or have one NULL value for one of those subs. 

We can consider a wrong implementation/bug if we have a situation like that.

Reply:
@Kingdutch 

> In my understanding, Drupal's constraint system is mostly there to help user input and runs only when validate is called on an entity. This means that programmatically we might still be able to create a field item that's invalid just because validation is bypassed. The question is then whether we want to allow this, or whether we want to prevent it and let a developer know "Hey you forgot to validate your inputs, that's not right!"

Yes we also had this discussion with @tbsiqueira and my argument here is: we should never let any entity being saved without being validated (except if you really know what you are doing and accept all consequences of doing it so). I also believe that validate method is overlooked many times, so I understand your concerns, but still `setValue` nor `isEmpty` methods as their name says, are not places to validate stuff (IMHO). 
If we really want to prevent errors if someone forget to do `->validate()` than we basically need to put all constraints into `setValue` method (and I am still not suggesting this :D). 

The whole business logic relies on correct data (ATM handled by constraints), so throwing an InvalidArgumentException if any of those subfields are not set is basically just one piece of warning if someone forget to validate, and this will still not prevent errors and application will still crash if invalid data will be saved without validation. :)

> "How do we warn a developer that they're trying to save an invalid field value because they (intentionally) skipped the validate step" and solve that problem.

Can we somehow do automatic check for this - so that this is rather part of code quality than how we build stuff :)

Reply:
> we should never let any entity being saved without being validated (except if you really know what you are doing and accept all consequences of doing it so). I also believe that validate method is overlooked many times, so I understand your concerns

I think in general that follows Drupal's own rules around validation. However, there's still valid cases where you want to bypass it (for example in one of our recent feature implementations for partial entities; or during migrations). So if we can help developers not forget the right thing (without eliminating their options) then we should try to do that.

> application will still crash if invalid data will be saved without validation

For me it's not necessarily about "preventing the application from crashing". If a developer implements a bug then that's one of the outcomes that may be unavoidable. However, our aim of exceptions should be that if we know a developer may make a mistake, we make it as easy as possible to know what that mistake was an how to fix them. So that's why I'm looking at "What happens if a developer doesn't validate and how difficult might it be to debug for them."

A contrived example would be something like:
```
/**
  * @param int[] $grades
  */
function averageGrades(array $grades) : float {
  return array_sum($grades) / count($grades);
}
```

We don't need to worry that a developer is passing a string because PHPStan will help us with that. However, in this case an empty array of grades would be invalid and would lead to the program crashing with a "division by 0" error, which might confuse the developer. To make debugging easier (because the division by 0 is not the actual problem) we can do:

```
/**
  * @param int[] $grades
  */
function averageGrades(array $grades) : float {
  if ($grades === []) {
     throw \InvalidArgumentException('An average can not be calculated without grades. $grades may not be an empty array.');
  }
  return array_sum($grades) / count($grades);
}
```

This has the same result if the developer does nothing: the program crashes with an exception. However, the information that the developer has is a lot easier to process.

--------

For me the rule of thumb that I try to use is: "Will not doing anything here have known errors down the line for certain inputs or will it cause unexpected behavior such as silently discarding a value". If the answer is "Yes" then we should try to capture those scenario's earlier and provide useful developer feedback. If the answer is "No" then we're all good and don't need to do anything else :)

Reply:
As discussed with @Kingdutch, in case developer save entity without `validate()` method, there is no good approach to inform the developer about mistake. In the setValue, we could check for constraints by calling `$this->validate()`, but if we would throw an exception we would break UI form. 

We also check how Drupal core is tackling this issue, but basically if you are not validating data, more or less, Drupal does not care and not trying prevent errors in this cases.

The only way (that I can think of now) if we need to throw an error for developer is to implement `hook_entity_presave` and trigger error there, by checking if validation was triggered or not - note via UI it will always trigger with validation. In case we will want to implement `hook_entity_presave` with described logic, we will do this in separated story/PR.

By this comment, I am resolving current conversation.